Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,330 @@
+-/
+def cons (x : α) (xs : Vector α n) : Vector α (n + 1) := Nat.add_comm 1 n ▸ (#v[x] ++ xs)
+
+/-- Make a vector of size `n` with all cells containing `v` -/
+def mkVector (n : Nat) (v : α) : Vector α n :=
+  ⟨mkArray n v, proof⟩
+  where
+    proof := by
+      rw [size_mkArray]
+
+/--
+`extract v start halt` Returns the slice of `v` from indices `start` to `stop` (exclusive). If `start` is greater or equal to `stop`, the result is empty. If `stop` is greater than the size of `v`, the size is used instead.
+-/
+def extract (v : Vector α n)
+  (start : Nat) (halt : Nat) : Vector α (min halt n - start) :=
+  ⟨Array.extract v.toArray start halt, proof⟩
+  where
+    proof := by
+      rw [size_extract, v.size_eq]
+
+/-- Custom eliminator for `Vector α n` through Arrays -/
+@[elab_as_elim]
+def elimAsArray {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : Array α, motive ⟨a, rfl⟩) :
+    {n : Nat} → (v : Vector α n) → motive v
+  | _, ⟨a, rfl⟩ => mk a
+
+/-- Custom eliminator for `Vector α n` throughLists -/
+@[elab_as_elim]
+def elimAsList {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : List α, motive ⟨⟨a⟩, rfl⟩) :
+    {n : Nat} → (v : Vector α n) → motive v
+  | _, ⟨⟨a⟩, rfl⟩ => mk a
+
+/-- Map a vector under a function. -/
+def map (f : α → β) : Vector α n → Vector β n
+  | ⟨a, h⟩ => ⟨Array.map f a, by simp [*]⟩
+
+/-- Mapping two vectors under a curried function of two variables. -/
+def map₂ (f : α → β → φ) : Vector α n → Vector β n → Vector φ n
+  | ⟨a, h₁⟩, ⟨b, h₂⟩ => ⟨Array.zipWith a b f, by simp [Array.size_zipWith, h₁, h₂]⟩
+
+/-- Drop `i` elements from a vector of length `n`; we can have `i > n`. -/
+def drop (i : Nat) (v : Vector α n) : Vector α (n - i) :=
+  have : min n n - i = n - i := by
+    rw [Nat.min_self]
+  this.symm ▸ (extract v i n)
+
+/-- Take `i` elements from a vector of length `n`; we can have `i > n`. -/
+def take (i : Nat) (v : Vector α n) : Vector α (min i n) :=
+  extract v 0 i
+
+/-- Vector of length `n` from a function on `Fin n`. -/
+def ofFn (f : Fin n → α) : Vector α n := ⟨Array.ofFn f, by {rw [size_ofFn]}⟩
+
+/-- Create a vector from another with a provably equal length. -/
+protected def cast {n m : Nat} (h : n = m) : Vector α n → Vector α m
+  | ⟨x, p⟩ => ⟨x, h ▸ p⟩
+
+/--
+Swaps two entries in an array, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that `v` has a reference
+count of 1 when called.
+-/
+def swap (v : Vector α n) (i j : Fin n) : Vector α n :=
+  ⟨Array.swap v.toArray (v.size_eq.symm ▸ i) (v.size_eq.symm ▸ j), proof⟩
+  where
+    proof := by
+      rw [size_swap]
+      exact v.size_eq
+
+/--
+Swaps two entries in an vector, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swap! (v : Vector α n) (i j : Nat) : Vector α n :=
+  ⟨Array.swap! v.toArray i j, by {rw [size_swap!]; exact v.size_eq}⟩
```

Comment:
Shouldn't we have a version of this that also takes indices as `Nat`, but that tries to prove that they are within the array bounds using `get_elem_tactic`?

Reply:
Doesn't it use omega now?

Reply:
How would it come up with such a proof without any statically provided hypothesis?

Reply:
More specifically, I need to provide this proof statically because I have to construct a Vector with a proof that the index `i` is less than `n`. `Array` doesn't have to be constructed with such a proof.

Reply:
Depends on the context in which it's used (for instance, by being in a `Std.Range`, as you say). All you need to do to write the function is to have a version that takes the proof as an additional parameter, with default value `by get_elem_tactic`.

Reply:
I tried it and it actually worked. What's a good suffix for these Nat indexed functions? I will add a Nat variant for every Fin index function that doesn't have a counterpart with a Nat index.

Reply:
I added a `swapAtNat` and `swapNat` function. But I am not happy with the names. Is there a better nomenclature for these functions?

Reply:
I'm not too sure about the names either. We have names like `getD` for the version where there's a default value; maybe `swapN`...?

Reply:
`swapN` sounds good to me.

Reply:
`swapN` also sounds good.

Reply:
Done. We now have `getN`, 'swapN`, `swapAtN` and `eraseIdxN`

Reply:
we also have `setN` now.